src/ir/auth_logic/lowering_ast_datalog.h

```diff
@@ -0,0 +1,450 @@
+  }
+
+  // In the same way that attributes are passed around with "CanActAs", so
+  // are other "CanActAs" facts. To implement this, the translation of
+  // CanActAs also results in both a predicate and an extra rule that passes
+  // these facts around. This function is for generating the extra rule and
+  // it works similarly to "SpokenAttributeToDLIR".
+  DLIRAssertion SpokenCanActAsToDLIR(Principal speaker, CanActAs can_act_as) {
+    Predicate main_predicate = CanActAsToDLIR(can_act_as);
+
+    // "canActAs" facts are passed to principals via other canActAs facts in
+    // essentially the same way as attributes. This function adds extra
+    // rules to pass these around. If the `canActAs` under translation
+    // is `X canActAs Z`, then the rule we need to generate is:
+    // `speaker says Y PredX :-
+    //    speaker says Y canActAs X, speaker says X canActAsZ`
+    // (Where Y is a fresh variable)
+    Principal prin_y = Principal(FreshVar());
+
+    // This is `speaker says Y PredX`
+    Predicate generated_lhs = PushPrin("says_", prin_y, main_predicate);
+
+    Predicate y_can_act_as_x = Predicate(
+        "canActAs", {prin_y.name(), can_act_as.left_principal().name()},
+        Sign::kPositive);
+
+    Predicate speaker_says_y_can_act_as_x =
+        PushPrin("says_", speaker, y_can_act_as_x);
+
+    // This is `speaker says X canActAs Z`
+    Predicate speaker_says_x_can_act_as_z =
+        PushPrin("says_", speaker, main_predicate);
+
+    // This is the full generated rule:
+    // `speaker says Y PredX :-
+    //    speaker says Y canActAs X, speaker says X canActAs Z`
+    return DLIRAssertion(DLIRCondAssertion(
+        generated_lhs, {std::move(speaker_says_y_can_act_as_x),
+                        std::move(speaker_says_x_can_act_as_z)}));
+  }
+
+  std::pair<Predicate, std::vector<DLIRAssertion>> BaseFactToDLIRInner(
+      Principal speaker, Predicate predicate) {
+    std::vector<DLIRAssertion> pred_list = {};
+    return std::make_pair(predicate, pred_list);
+  }
+
+  std::pair<Predicate, std::vector<DLIRAssertion>> BaseFactToDLIRInner(
+      Principal speaker, Attribute attribute) {
+    std::vector<DLIRAssertion> pred_list = {
+        SpokenAttributeToDLIR(speaker, attribute)};
+    return std::make_pair(AttributeToDLIR(attribute), pred_list);
+  }
+
+  std::pair<Predicate, std::vector<DLIRAssertion>> BaseFactToDLIRInner(
+      Principal speaker, CanActAs canActAs) {
+    std::vector<DLIRAssertion> pred_list = {
+        SpokenCanActAsToDLIR(speaker, canActAs)};
+    return std::make_pair(CanActAsToDLIR(canActAs), pred_list);
+  }
+
+  // The second return value represents 0 or 1 newly generated rules, so an
+  // option might seem more intuitive. However, the interface that consumes
+  // this needs to construct a vector anyway, so a vector is used in the
+  // return type.
+  std::pair<Predicate, std::vector<DLIRAssertion>> BaseFactToDLIR(
+      Principal speaker, BaseFact base_fact) {
```

Comment:
Can you change the arguments of all functions (except constructors)  to pass by const reference?  i.e., `const Principal&` instead of `Principal`, `const BaseFact&` instead of `BaseFact`. 

Without `std::move`, we seem to be making too many copies.

Reply:
Yes. This makes sense to me.